### ISSUE TITLE
Adding a new "single" type of data (in addition to current "array")

### ISSUE THREAD

**Issue by marcgarba** [rsmp-nordic/rsmp_core#133]
If in a SXL we want to gather somes informations (defined in it as a whole for a CodeId/name) instead of using separated return names (or commands),
we can use the "array" type.

For its many advantages:
- more compact return exchange,
- possible optional information,
- YAML definition (instead of "free" string with specific encoding described as a 'comment'),
- no bad linked data synchronisation risk (with individual return name status subscribtion),
- ...

But... the type "array" is normally planned to contains a list of differents elements (list of JSON objects with "[....]" format).
In this case (informations of an object that goes together), only "one" object is referenced ("cId"), and only and always one JSON object element will be put in the array.

It would be more suitable to use a new "single" data type added to RSMP for this SXL exchange case.
The string will just contains a JSON object.
Example for 's' : 
`[{"PropA":"1","PropB":"1457","PropC":"2.3"}]`
that becomes : 
`{"PropA":"1","PropB":"1457","PropC":"2.3"}`

This would allow strict control for JSON schema: no more risk to have abnormal multiples elements in a array, if not an array !
For the roadside simulator, when enquiring values, we can differentiate the case "single" / "array" type, and if a "single" directly open the item second window (without the previous items list manager first window, only useful for an array of items).

**Comment by emiltin** [comment]
thanks! i think this is similar to https://github.com/rsmp-nordic/rsmp_core/issues/84

**Comment by marcgarba** [comment]
No, the request for a new "single" type, is just a specific case of the "array", when always only one element in, present.
Absolutely no impact on RSMP 'core'.

It as always difficult to give an example to illustrate,
but if we have for example a signal CodeID "S123" / name "thesize", and we want to get width/height data values (together) with it:
currently with the array type format, we will obtain :
`"[{"w":"12.3","h":"4.5"}]"`
as there is always only "one" size, it would have been preferably to obtain for this specific case data (limited to one item) : 
`"{"w":"12.3","h":"4.5"}"`

For the roadside simulator, this would make it possible to enter directly and only the 2 width/height values, without managing any list of elements, which have no meaning here...

Hope it helps to understand better the need for this little addition...
Thanks.

**Comment by emiltin** [comment]
Hi,
Usually you would not pack several items into into a single SXL attribute, because
1. you loose the ability to request and subscribe to just the part you need
2. you need to send everything each time just one item changes

But for the example of sending a dimension that consists of width and height, it could make sense to pack these together. The best format would be to use a json map (hash) like you suggest:

```json
{"w":"12.3", "h":"4.5"}
```

RSMP core spec does not currently describe maps as a type (like we do with arrays), but I think it would make sense to allow any json data, also nested types. In general, we're moving towards using native json types everywhere. I would also prefer numbers to be send a json types, rather than strings, so for your example:

```json
{"w":12.3, "h":4.5}
```

Data placed inside strings is much harder to validate with json schema (often unfeasible).

**Comment by emiltin** [comment]
An quick solution using the current arrays would be to send it as:
```json
["12.3","4.5"]
````
Defining first item as width, second as height.

**Comment by marcgarba** [comment]
Thanks for your reply,
and good if with the little "size" example you seen the interest to introduce this new type in RSMP.

About the format for numbers in this new type, it was just using the same format than actually in "array".
Example with current "S0033" of TLC :
```
[{"r": "uhnv","t": "2021-11-09T15:04:12.348Z","s": "completed","e": "5","d": "10"},
{"r": "oh0i","t": "2021-11-09T15:06:38.796Z","s": "activated"}]
```
So in the future, as it would be logic to have the same format for numbers in theses 2 types, it would also become :
```
[{"r": "uhnv","t": "2021-11-09T15:04:12.348Z","s": "completed","e": 5,"d": 10},
{"r": "oh0i","t": "2021-11-09T15:06:38.796Z","s": "activated"}]
```
Planned like that ?
For now not a lot of signals using the recent type 'array", so perhaps changing it fastly would be better...

**Comment by emiltin** [comment]
In general yes, we want to move towards native json types for numbers, booleans, arrays and maps (also known as json objects), null. And of course we would still have strings.  Note that json does not include a native type for dates.

**Comment by emiltin** [comment]
Related to #84

**Comment by marcgarba** [comment]
No, as already explained previously, the request for a new "single" type, is just a specific case of the "array", when always only one element in, present.
_Absolutely no impact on RSMP 'core'._

In fact, this issue is now detailed in #135 : **"Allow JSON Objects (maps)"**

**Comment by marcgarba** [comment]
This issue has now been processed, and **shoud be closed** !
and one less...
